Issue 4758 - Add tests for WebUI

bsimonova opened the pull request:
Description:
Added some basic tests for WebUI (login, create instance etc.). These tests need
specific environment to be run so they will be skipped in normal conditions.
Playwright for Python was used to develop the tests and they pass in latest
RHDS version with Firefox and Chromium.

Relates: https://github.com/389ds/389-ds-base/issues/4758

Reviewed by: ???

vashirov commented on dirsrvtests/tests/suites/webui/__init__.py:
```diff
@@ -0,0 +1,62 @@
+    page.fill('#login-user-input', 'root')
+    page.fill('#login-password-input', 'redhat')
+    page.click("#login-button")
+    page.wait_for_selector('text=Red Hat Directory Server')
```
I've added missing dependencies. Test now fails here. On Fedora it's "389 Directory Server".
It would be also nice to capture screenshot on failure, so it would be easier to investigate such failures. See example here https://github.com/microsoft/playwright-pytest#screenshot-on-test-failure

vashirov replied:
And `DEBUG=pw:api` would be useful too for stack traces from playwright.

bsimonova replied:
Ahh, I knew I wanted to include screenshots :D thanks for reminder.
I had to install slugify and unicode-slugify to make the paths work. Would this dependency be ok with us or do we want to modify it without slugify?
You mean to add DEBUG=pw:api to pytest.yml?

vashirov replied:
> I had to install slugify and unicode-slugify to make the paths work. Would this dependency be ok with us or do we want to modify it without slugify?

Are both of these needed or just `python-slugify`?

> You mean to add DEBUG=pw:api to pytest.yml?

Correct.

bsimonova replied:
Just python-slugify should be enough.

vashirov commented on dirsrvtests/tests/suites/webui/__init__.py:
```diff
@@ -0,0 +1,62 @@
+
+    # We are at login page
+    page.fill('#login-user-input', 'root')
+    page.fill('#login-password-input', 'redhat')
```
Found another issue with the CI container image. We don't have password set there. I don't like the idea of hardcoding the password, I'd rather generate it on runtime, change it before the start of the test and pass the value to the test, so it can use it. I'll send you the proposed changes.

bsimonova replied:
I also don't like the hardcoded value so I am glad there is another suggestion. Thanks, will check the changes.

vashirov commented on dirsrvtests/tests/suites/webui/login/login_test.py:
```diff
@@ -0,0 +1,113 @@
+    assert page.url == "http://localhost:9090/389-console"
+
+    log.info('Let us log out')
+    page.click('#content-user-name')
```
In cockpit-250 on F34 it's a new selector "#navbar-dropdown".

vashirov commented on dirsrvtests/tests/suites/webui/__init__.py:
```diff
@@ -0,0 +1,68 @@
+        page.wait_for_selector('text=Red Hat Directory Server')
+        page.click('text=Red Hat Directory Server')
+    else:
+        page.wait_for_selector('389 Directory Server')
```
Missing `text=` in the selector

vashirov commented on .github/workflows/pytest.yml:
```diff
@@ -65,7 +65,9 @@ jobs:
-        sudo docker exec $CID py.test  --suppress-no-test-exit-code  -m "not flaky" --junit-xml=pytest.xml -v dirsrvtests/tests/suites/${{ matrix.suite }}
+        export PASSWD=$(openssl rand -base64 32)
+        sudo docker exec $CID sh -c "echo \"${PASSWD}\" | passwd --stdin root"
+        sudo docker exec -e WEBUI=1 -e DEBUG=pw:api -e PASSWD="${PASSWD}" $CID py.test  --suppress-no-test-exit-code  -m "not flaky" --junit-xml=pytest.xml -v dirsrvtests/tests/suites/${{ matrix.suite }}
```
We should also start `cockpit.socket`, but maybe I should fix it in the image itself.

vashirov commented on dirsrvtests/tests/suites/webui/create/create_instance_test.py:
```diff
@@ -0,0 +1,212 @@
+         2. Button is visible
+    """
+
+    frame = page.frame('cockpit1:localhost/389-console')
```
All tests fail with firefox, because for some reason selection of the frame by name returns `None`. If I select a frame from `page.frames` list, then it works. But the index of the required frame can change. 
I ran codegen to see what it suggests, and after adding this snippet, frame selection works:
```
    # Click text=389 Directory Server
    # with page.expect_navigation(url="http://localhost:9090/389-console"):
    with page.expect_navigation():
        page.click("text=389 Directory Server")
    assert page.url == "http://localhost:9090/389-console"

```

bsimonova replied:
I applied the changes and we should have the tests fixed for chromium now at least. But for firefox I still have issues accessing the frame. Your snippet works but it fails with frames when proceeding to create instance.
I ran codegen and it produced this:
```
         # Click text=389 Directory Server
         # with page.expect_navigation(url="https://localhost:9090/389-console"):
         with page.expect_navigation():
             page.click("text=389 Directory Server")
         # assert page.url == "http://localhost:9090/389-console"

         # Go to https://localhost:9090/389-console
         page.goto("http://localhost:9090/389-console")

         # Click text=Create New Instance
         page.frame(name="cockpit1:localhost/389-console").click("text=Create New Instance")
```

The last line works in codegen but fails in tests again with a NoneType object. I have to investigate more.

bsimonova replied:
@vashirov
Also I see it fails in checks again with connection refused. All  tests pass on my machine for chromium (I am still fixing firefox). Could it be an issue with how I set the password for the docker?

vashirov replied:
IIRC there was an issue with `dbus.service` and `cockpit.socket` not starting by default. 
Could you please add this before running the tests to see if it helps:
```
sudo docker exec $CID sh -c "systemctl start dbus.service"
sudo docker exec $CID sh -c "systemctl enable --now cockpit.socket"
```

vashirov replied:
I've added cockpit to the test image and enabling `cockpit.socket` works now. Some tests have finally passed, but some are failing due to timeout. There should be screenshots generated on failure, but currently they are not captured as part of the artifact collection step, we need to extend it too. I'll look into this tomorrow.

bsimonova committed:
Issue 4758 - Add tests for WebUI

Description:
Added some basic tests for WebUI (login, create instance etc.). These tests need
specific environment to be run so they will be skipped in normal conditions.
Playwright for Python is used to develop the tests and they can run
by using Firefox and Chromium browser.

Relates: https://github.com/389ds/389-ds-base/issues/4758

Reviewed by: @vashirov (Thanks!)

vashirov reviewed (approved):
Looks good, let's merge this :)
